pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,858 @@
+		AddFunc:    c.addPacketCapture,
+		UpdateFunc: c.updatePacketCapture,
+		DeleteFunc: c.deletePacketCapture,
+	}, resyncPeriod)
+
+	c.newPacketSourceFn = func(c *Controller, pc *crdv1alpha1.PacketCapture) (PacketSource, error) {
+		matchPacket, err := c.createMatchPacket(pc)
+		if err != nil {
+			return nil, err
+		}
+		device := c.getTargetCaptureDevice(pc)
+		if device == nil {
+			return nil, nil
+		}
+		filter := bpf.CompilePacketFilter(pc.Spec.Packet, matchPacket)
+		klog.V(5).InfoS("Generated bpf instructions for Packetcapture", "name", pc.Name, "match packet", matchPacket, "bpf instructions", filter)
+		return NewPcapSource(*device, filter)
+	}
+
+	return c
+}
+
+func (c *Controller) enqueuePacketCapture(pc *crdv1alpha1.PacketCapture) {
+	c.queue.Add(pc.Name)
+}
+
+// Run will create defaultWorkers workers (go routines) which will process the PacketCapture events from the
+// workqueue.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting packetcapture controller", "name", controllerName)
+	defer klog.InfoS("Shutting down packetcapture controller", "name", controllerName)
+
+	cacheSynced := []cache.InformerSynced{c.packetCaptureSynced}
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSynced...) {
+		return
+	}
+
+	err := defaultFS.MkdirAll(packetDirectory, 0755)
+	if err != nil {
+		klog.ErrorS(err, "Couldn't create the directory for storing captured packets", "directory", packetDirectory)
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *Controller) addPacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture ADD event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) updatePacketCapture(_, obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture UPDATE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func (c *Controller) deletePacketCapture(obj interface{}) {
+	pc := obj.(*crdv1alpha1.PacketCapture)
+	klog.InfoS("Processing PacketCapture DELETE event", "name", pc.Name)
+	c.enqueuePacketCapture(pc)
+}
+
+func nameToPath(name string) string {
+	return filepath.Join(packetDirectory, name+".pcapng")
+}
+
+func (c *Controller) worker() {
+	// check the captures that are waiting in line.
+	go c.processWaitingCaptures()
+	for c.processPacketCaptureItem() {
+	}
+}
+
+func (c *Controller) processWaitingCaptures() {
+	for {
+		capture := func() string {
+			c.cond.L.Lock()
+			defer c.cond.L.Unlock()
+			for c.numRunningCaptures >= maxConcurrentCaptures || c.waitingCaptures == 0 {
+				c.cond.Wait()
+			}
+			c.numRunningCaptures += 1
+			c.waitingCaptures -= 1
+			capture := <-c.waitingCapturesCh
+			c.startPacketCapture(capture)
+			c.captures[capture] = packetCapturePhaseRunning
+			return capture
+		}()
+		c.queue.Add(capture)
```

Comment:
I don't quite understand why `startPacketCapture` could be called here and in `syncPacketCapture`. 
Why can't it use `syncPacketCapture` to handle this? Having another goroutine to start capture makes the concurrency hard to understand and error-prone.

Reply:
`syncPacketCapture` and `processWaitingCaptures` handle totally different concurrency cases and then called `startPacketCapture`.  I think separate them was to make the condition check easier to understand and move all the actual capture work to `startPacketCapture`

Reply:
My point is to avoid different concurrency cases in a single controller, which makes code complicated and error-prone. In another PR under review (not yours), we have to use locks everywhere to avoid race condition after introducing a new concurrency to handle an edge case. The concurrency in that PR is inevitable, but I don't see it necessary in this PR, and it does involve various synchronous mechanism and data structures to implement it, which should be avoided as much as possible.

My question is, why we can't just use the workers to handle all jobs of a capture request? I imagine the worker's logic is as follows:
1. If the capture hasn't started,
1.a. if the maximum concurrency is reached, put the item in a waiting list (or just enqueue it with a delay in this release to make it simple), return.
1.b. if the maximum concurrency is not reached, start capturing in a goroutine, update status, return.
2. If the capture has started,
2.a. if the required number is reached, upload the pcap, update status, return.
2.b. if the required number is not reached, update status, return.
3. If the capture is not found, remove the file.